Tries to install html-webpack-plugin-loader if ^2.21.0

If *html-webpack-plugin* ^2.21.0 is installed (tested with 2.26.0), *npm-install-webpack-plugin* tries to install it as a loader. Interestingly 2.21.0 works just fine. A release after that broke the behavior as it gets detected improperly.

To make this easier to debug, I set up a [repository](https://github.com/survivejs/react-boilerplate/tree/npm-install-webpack-plugin-bug). `npm install` and `npm start` to get

```bash
Installing html-webpack-plugin-loader...
npm ERR! Darwin 16.3.0
npm ERR! argv "/usr/local/Cellar/node6-lts/6.9.1/bin/node" "/usr/local/bin/npm" "install" "html-webpack-plugin-loader" "--save"
npm ERR! node v6.9.1
npm ERR! npm  v3.10.8
npm ERR! code E404

npm ERR! 404 Registry returned 404 for GET on https://registry.npmjs.org/html-webpack-plugin-loader
npm ERR! 404
npm ERR! 404  'html-webpack-plugin-loader' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.
```

--- a/example/webpack2/.babelrc
+++ b/example/webpack2/.babelrc
@@ -4,7 +4,7 @@
   ],
   "presets": [
     "react",
-    "es2015-webpack",
+    ["es2015", {"modules": false}],
     "stage-0"
   ],
   "env": {

--- a/example/webpack2/package.json
+++ b/example/webpack2/package.json
@@ -2,7 +2,7 @@
   "private": true,
   "devDependencies": {
     "reload-server-webpack-plugin": "1.0.1",
-    "webpack": "2.1.0-beta.6"
+    "webpack": "2.2.0"
   },
   "scripts": {
     "link": "(cd .. && npm link .) && npm link npm-install-webpack-plugin",

--- a/example/webpack2/src/client.js
+++ b/example/webpack2/src/client.js
@@ -2,6 +2,6 @@
 import App from "App";
 
 // Testing inline-loaders
-import loadRender from "bundle?lazy!./render";
+import loadRender from "bundle-loader?lazy!./render";
 
 loadRender(({ render }) => render(App));

--- a/example/webpack2/webpack.config.client.js
+++ b/example/webpack2/webpack.config.client.js
@@ -12,7 +12,7 @@ module.exports = Object.assign({}, defaults, {
   },
 
   output: Object.assign({}, defaults.output, {
-    libaryTarget: "var",
+    libraryTarget: "var",
     path: path.join(defaults.context, "build/client"),
     publicPath: "/",
   }),

--- a/example/webpack2/webpack.config.defaults.js
+++ b/example/webpack2/webpack.config.defaults.js
@@ -8,15 +8,14 @@ module.exports = {
 
   module: {
     loaders: [
-      { test: /\.css$/, loader: "style" },
-      { test: /\.css$/, loader: "css", query: { localIdentName: "[name]-[local]--[hash:base64:5]" } },
-      { test: /\.eot$/, loader: "file" },
-      { test: /\.js$/, loader: "babel", query: { cacheDirectory: true }, exclude: /node_modules/ },
-      { test: /\.json$/, loader: "json" },
-      { test: /\.(png|jpg)$/, loader: "url", query: { limit: 8192 } }, // Inline base64 URLs for <= 8K images
-      { test: /\.svg$/, loader: "url", query: { mimetype: "image/svg+xml" } },
-      { test: /\.ttf$/, loader: "url", query: { mimetype: "application/octet-stream" } },
-      { test: /\.(woff|woff2)$/, loader: "url", query: { mimetype: "application/font-woff" } },
+      { test: /\.css$/, loader: "style-loader" },
+      { test: /\.css$/, loader: "css-loader", options: { localIdentName: "[name]-[local]--[hash:base64:5]" } },
+      { test: /\.eot$/, loader: "file-loader" },
+      { test: /\.js$/, loader: "babel-loader", options: { cacheDirectory: true }, exclude: /node_modules/ },
+      { test: /\.(png|jpg)$/, loader: "url-loader", options: { limit: 8192 } }, // Inline base64 URLs for <= 8K images
+      { test: /\.svg$/, loader: "url-loader", options: { mimetype: "image/svg+xml" } },
+      { test: /\.ttf$/, loader: "url-loader", options: { mimetype: "application/octet-stream" } },
+      { test: /\.(woff|woff2)$/, loader: "url-loader", options: { mimetype: "application/font-woff" } },
     ],
   },
 

--- a/package.json
+++ b/package.json
@@ -16,6 +16,7 @@
     "coverage": "npm test && nyc report --reporter=text-lcov | coveralls",
     "postversion": "npm run version:amend && git push origin master --tags && npm publish",
     "test": "cross-env NODE_ENV=test nyc mocha",
+    "test:watch": "cross-env NODE_ENV=test mocha --watch",
     "version": "npm run changelog",
     "version:amend": "git commit --amend -m \"Release v${npm_package_version}\""
   },

--- a/src/installer.js
+++ b/src/installer.js
@@ -9,6 +9,17 @@ var PEERS = /UNMET PEER DEPENDENCY ([a-z\-0-9\.]+)@(.+)/gm;
 var defaultOptions = { dev: false, peerDependencies: true };
 var erroneous = [];
 
+function normalizeBabelPlugin(plugin, prefix) {
+  // Babel plugins can be configured as [plugin, options]
+  if (Array.isArray(plugin)) {
+    plugin = plugin[0];
+  }
+  if (plugin.indexOf(prefix) === 0) {
+    return plugin;
+  }
+  return prefix + plugin;
+}
+
 module.exports.check = function(request) {
   if (!request) {
     return;
@@ -99,13 +110,13 @@ module.exports.checkBabel = function checkBabel() {
 
   // Accumulate babel-core (required for babel-loader)+ all dependencies
   var deps = ["babel-core"].concat(options.plugins.map(function(plugin) {
-    return "babel-plugin-" + plugin;
+    return normalizeBabelPlugin(plugin, "babel-plugin-");
   })).concat(options.presets.map(function(preset) {
-    return "babel-preset-" + preset;
+    return normalizeBabelPlugin(preset, "babel-preset-");
   })).concat(options.env.development.plugins.map(function(plugin) {
-    return "babel-plugin-" + plugin;
+    return normalizeBabelPlugin(plugin, "babel-plugin-");
   })).concat(options.env.development.presets.map(function(preset) {
-    return "babel-preset-" + preset;
+    return normalizeBabelPlugin(preset, "babel-preset-");
   }));
 
   // Check for missing dependencies

--- a/src/plugin.js
+++ b/src/plugin.js
@@ -2,6 +2,7 @@ var MemoryFS = require("memory-fs");
 var webpack = require("webpack");
 
 var installer = require("./installer");
+var utils = require("./utils");
 
 var depFromErr = function(err) {
   if (!err) {
@@ -115,7 +116,7 @@ NpmInstallPlugin.prototype.resolveExternal = function(context, request, callback
     request: request,
   };
 
-  this.resolve(result, function(err, filepath) {
+  this.resolve('normal', result, function(err, filepath) {
     if (err) {
       this.install(Object.assign({}, result, { request: depFromErr(err) }));
     }
@@ -124,12 +125,12 @@ NpmInstallPlugin.prototype.resolveExternal = function(context, request, callback
   }.bind(this));
 };
 
-NpmInstallPlugin.prototype.resolve = function(result, callback) {
+NpmInstallPlugin.prototype.resolve = function(resolver, result, callback) {
   var version = require("webpack/package.json").version;
   var major = version.split(".").shift();
 
   if (major === "1") {
-    return this.compiler.resolvers.normal.resolve(
+    return this.compiler.resolvers[resolver].resolve(
       result.path,
       result.request,
       callback
@@ -137,7 +138,7 @@ NpmInstallPlugin.prototype.resolve = function(result, callback) {
   }
 
   if (major === "2") {
-    return this.compiler.resolvers.normal.resolve(
+    return this.compiler.resolvers[resolver].resolve(
       result.context || {},
       result.path,
       result.request,
@@ -145,23 +146,31 @@ NpmInstallPlugin.prototype.resolve = function(result, callback) {
     );
   }
 
-  throw new Error("Unsupport Webpack version: " + version);
+  throw new Error("Unsupported Webpack version: " + version);
 }
 
 NpmInstallPlugin.prototype.resolveLoader = function(result, next) {
-  // Ensure loaders end with `-loader` (e.g. `babel` => `babel-loader`)
-  // Also force Webpack2's duplication of `-loader` to a single occurrence
-  var loader = result.request // e.g. react-hot-loader/webpack
-      .split("/")             // ["react-hot-loader", "webpack"]
-      .shift()                // "react-hot-loader"
-      .split("-loader")       // ["react-hot", ""]
-      .shift()                // "react-hot"
-      .concat("-loader")       // "react-hot-loader"
-  ;
+  // Only install direct dependencies, not sub-dependencies
+  if (result.path.match("node_modules")) {
+    return next();
+  }
 
-  this.install(Object.assign({}, result, { request: loader }));
+  if (this.resolving[result.request]) {
+    return next();
+  }
 
-  return next();
+  this.resolving[result.request] = true;
+
+  this.resolve("loader", result, function(err, filepath) {
+    this.resolving[result.request] = false;
+
+    if (err) {
+      var loader = utils.normalizeLoader(result.request);
+      this.install(Object.assign({}, result, { request: loader }));
+    }
+
+    return next();
+  }.bind(this));
 };
 
 NpmInstallPlugin.prototype.resolveModule = function(result, next) {
@@ -176,7 +185,7 @@ NpmInstallPlugin.prototype.resolveModule = function(result, next) {
 
   this.resolving[result.request] = true;
 
-  this.resolve(result, function(err, filepath) {
+  this.resolve('normal', result, function(err, filepath) {
     this.resolving[result.request] = false;
 
     if (err) {

--- a/src/utils.js
+++ b/src/utils.js
@@ -1,0 +1,12 @@
+/**
+ * Ensure loaders end with `-loader` (e.g. `babel` => `babel-loader`)
+ * Also force Webpack2's duplication of `-loader` to a single occurrence
+ */
+module.exports.normalizeLoader = function(loader) {
+  return loader             // e.g. react-hot-loader/webpack
+         .split("/")        // ["react-hot-loader", "webpack"]
+         .shift()           // "react-hot-loader"
+         .split("-loader")  // ["react-hot", ""]
+         .shift()           // "react-hot"
+         .concat("-loader") // "react-hot-loader"
+}

--- a/test/installer.test.js
+++ b/test/installer.test.js
@@ -149,7 +149,7 @@ describe("installer", function() {
 
     context("when .babelrc exists", function() {
       beforeEach(function() {
-        process.chdir(path.join(process.cwd(), "example/webpack1"));
+        process.chdir(path.join(process.cwd(), "example/webpack2"));
 
         this.check = expect.spyOn(installer, "check").andCall(function(dep) {
           return dep;

--- a/test/plugin.test.js
+++ b/test/plugin.test.js
@@ -116,7 +116,7 @@ describe("plugin", function() {
 
   describe(".resolveExternal", function() {
     beforeEach(function() {
-      this.resolve = expect.spyOn(this.plugin, "resolve").andCall(function(result, callback) {
+      this.resolve = expect.spyOn(this.plugin, "resolve").andCall(function(resolver, result, callback) {
         callback(new Error(util.format("Can't resolve '%s' in '%s'", result.request, result.path)));
       });
     });
@@ -153,56 +153,37 @@ describe("plugin", function() {
   });
 
   describe(".resolveLoader", function() {
-    it("should call installer.check", function() {
-      var result = { path: "node_modules", request: "babel-loader" };
+    it("should call .resolve", function() {
+      var result = { path: "/", request: "babel-loader" };
+
+      this.compiler.resolvers.loader.resolve.andCall(function(context, path, request, callback) {
+        callback(null);
+      }.bind(this));
+
+      var install = expect.spyOn(this.plugin, 'install')
 
       this.plugin.resolveLoader(result, this.next);
 
-      expect(this.check.calls.length).toBe(1);
-      expect(this.check.calls[0].arguments).toEqual(["babel-loader"]);
+      expect(this.compiler.resolvers.loader.resolve.calls.length).toBe(1);
+      expect(install.calls.length).toBe(0)
+      expect(this.next.calls.length).toBe(1);
+      expect(this.next.calls[0].arguments).toEqual([]);
     });
+    it("should call .resolve and install if not resolved", function() {
+      var result = { path: "/", request: "babel-loader" };
 
-    it("should skip installer.install if existing", function() {
-      var result = { path: "node_modules", request: "babel-loader" };
+      this.compiler.resolvers.loader.resolve.andCall(function(context, path, request, callback) {
+        callback(new Error("Can't resolve 'babel-loader' in 'node_modules'"));
+      }.bind(this));
 
-      this.check.andCall(function(dep) {
-        return false;
-      });
-
-      this.plugin.resolveLoader(result, this.next);
-
-      expect(this.install.calls.length).toBe(0);
-    });
-
-    it("should call installer.install if missing", function() {
-      var result = { path: "node_modules", request: "babel-loader" };
-
-      this.check.andCall(function(dep) {
-        return dep;
-      });
+      var install = expect.spyOn(this.plugin, 'install')
 
       this.plugin.resolveLoader(result, this.next);
 
-      expect(this.install.calls.length).toBe(1);
-      expect(this.install.calls[0].arguments).toEqual(["babel-loader", this.options]);
-    });
-
-    it("should convert `babel` to `babel-loader`", function() {
-      var result = { path: "node_modules", request: "babel" };
-
-      this.plugin.resolveLoader(result, this.next);
-
-      expect(this.check.calls.length).toBe(1);
-      expect(this.check.calls[0].arguments).toEqual(["babel-loader"]);
-    });
-
-    it("should convert ignore `react-hot-loader/webpack`", function() {
-      var result = { path: "node_modules", request: "react-hot-loader/webpack" };
-
-      this.plugin.resolveLoader(result, this.next);
-
-      expect(this.check.calls.length).toBe(1);
-      expect(this.check.calls[0].arguments).toEqual(["react-hot-loader"]);
+      expect(this.compiler.resolvers.loader.resolve.calls.length).toBe(1);
+      expect(install.calls.length).toBe(1)
+      expect(this.next.calls.length).toBe(1);
+      expect(this.next.calls[0].arguments).toEqual([]);
     });
   });
 

--- a/test/utils.test.js
+++ b/test/utils.test.js
@@ -1,0 +1,15 @@
+var expect = require("expect");
+
+var utils = require("../src/utils");
+
+describe("utils", function() {
+  describe(".normalizeLoader", function() {
+    it("should convert `babel` to `babel-loader`", function() {
+      expect(utils.normalizeLoader("babel")).toBe("babel-loader");
+    });
+
+    it("should convert `react-hot-loader/webpack` to `react-hot-loader`", function() {
+      expect(utils.normalizeLoader("react-hot-loader/webpack")).toBe("react-hot-loader");
+    });
+  })
+})

